### DISCUSSION TITLE
Raspberry Pico (thumbv6m-none-eabi) compatibility

### DISCUSSION THREAD

**Discussion by mangoplane** [Infineon/optiga-trust-m#101] [Q&A] [answered]
Hi,

I'm thinking of purchasing this hardware random number generator (RNG) device for a project I'm working on with Raspberry Pi Pico. Can you confirm if the product works with Raspberry Pi Pico and if it can be utilized through Rust's hardware abstraction layer?

How difficult would this be? Skimming over your examples, it does seem rather complicated to someone new to hardware like me.

Thanks.

**Comment by mangoplane** [comment]
Hi there, it appears that using the hardware RNG device with Raspberry Pi Pico is possible according to the information found in this resource: https://learn.adafruit.com/trinkey-qt2040-true-random-number-generator/sending-trng.

Additionally, this tutorial (https://www.youtube.com/watch?v=zSWkrpu8KBA) demonstrates that embedded C and embedded Rust through the Hardware Abstraction Layer (HAL) can coexist. The challenge may lie in integrating the library with Rust. Any assistance or guidance on this would be greatly appreciated.

**Reply by mangoplane** [reply]
I have watched the tutorial and I understand that it is possible to call functions written in Rust from C using extern/linking and vice versa. Based on this information, my plan is to write the majority of the hardware logic in C/C++, where I can utilize the Arduino library for the secure element True Random Number Generator (TRNG). However, for parts of the logic where accuracy is crucial, I will rely on functions implemented in Rust. For example, I will pass the TRNG values to Rust through an extern function, where they will be used.

**Reply by mangoplane** [reply]
For clarity and for the benefit of future readers, I would like to note that the vast majority of libraries designed for modern 32-bit CPU-based Arduino boards can be utilized with the RPI Pico. This includes the library specifically designed for this device. For more information, please refer to the following link: https://arduino-pico.readthedocs.io/en/latest/libraries.html

**Comment by ShaikKarishma0** [comment, marked as answer]
Hi @coregrad,
The above approach could be possible. If you're facing any issue with respect to OPTIGA Trust M, then please summarize it to get relevant support.
